scripts/pylib/twister/twisterlib/runner.py

```diff
@@ -511,6 +546,71 @@ def process(self, pipeline, done, message, lock, results):
+        if os.path.isfile(symbol_file):
+            logger.debug(f"zephyr.symbols found: {symbol_file}")
+        else:
+            # No zephyr.symbols file, cannot do symbol-based test case collection
+            logger.debug(f"zephyr.symbols NOT found: {symbol_file}")
+            return
+
+        yaml_testsuite_name = self.instance.testsuite.id
+        logger.debug(f"Collecting and amending test cases for test suite: {yaml_testsuite_name}")
+
+        with open(symbol_file, 'r') as fp:
+            symbols = fp.read()
+            logger.debug(f"test instance {self.instance.name} already has {len(self.instance.testcases)} cases.")
+
+            # Test case amending logic
+            for testcase in self.instance.testcases:
+                # Identifier is "yaml_testsuite.test_function" or just "yaml_testsuite"
+                identifier = testcase.name
+                if identifier == yaml_testsuite_name:
+                    # Sometimes twsiter add yaml test suite name to a test instance as a test case.
+                    # Such test case has no real test function.
+                    # Skip it to make stats right
+                    continue
+
+                # For the old-style ztest function
+                test_func_name1 = f"test_{identifier.split('.')[-1]}"
+                # For the new-style ztest function
+                test_func_name2_regex = re.compile("z_ztest_unit_test_.*" + identifier.split('.')[-1])
+
+                # If the zephyr.symbols file exists and the test case function not in it,
+                # it should be an un-intended test function for this test instance build.
+                # If this is an old test plan and contains test cases collected at runtime
+                # by checking the log, such test cases may not have functions in zephyr.symbols file.
+                # Also amend them here, which should be OK because they will be added back again at runtime.
+                m = test_func_name2_regex.search(symbols)
+                if test_func_name1 not in symbols and not m:
+                    testcase.status = "skipped"
+                    testcase.reason = "Test case doesn't belong to the test scenario"
+                    # results.skipped_runtime += 1
+                    continue
+
+            # Test case collecting logic
+            # It is only meant for new ztest fx because only new ztest fx exposes test functions
+            # precisely.
+
+            # The 1st capture group is new ztest suite name.
+            # The 2nd capture group is new ztest unit test name.
+            new_ztest_unit_test_regex = re.compile(r"z_ztest_unit_test__([^\s]*)__([^\s]*)")
+            matches = new_ztest_unit_test_regex.findall(symbols)
+            for m in matches:
+                # new_ztest_suite = m[0] # not used for now
+                test_func_name = m[1].replace("test_", "")
+                testcase_id = f"{yaml_testsuite_name}.{test_func_name}"
```

Comment:
Can we include `new_ztest_suite` in the testcase_id?  I feel this info is lost in the organization of test report. So something like 
```
testcase_id = f"{yaml_test_config}.{ztest_suite}.{ztest_name}"
```

Reply:
I did think about this before.  Line 605 is the relic of my thought. 

If I add the test case including the test suite, like `{yaml_testsuite_name}.{new_ztest_suite}.{test_func_name}`, when the harness sees a test case in the runtime serial log with the name of `{yaml_testsuite_name}.{test_func_name}`, there's no ztest suite info in the log line, so it will still consider it as a missing test case. And add it using below logic.

https://github.com/zephyrproject-rtos/zephyr/blob/8c7eba346af3119f5920a288af93f0fa5efd76eb/scripts/pylib/twister/twisterlib/harness.py#L14

https://github.com/zephyrproject-rtos/zephyr/blob/8c7eba346af3119f5920a288af93f0fa5efd76eb/scripts/pylib/twister/twisterlib/harness.py#L273-L275

https://github.com/zephyrproject-rtos/zephyr/blob/1bbe209a30fcdab037c8ac9b96c33ea2249665e0/scripts/pylib/twister/twisterlib/testinstance.py#L111-L119


This will cause incorrect stats.  And the newly added one will still be considered as `No status`.


For now I prefer to leave the `new_ztest_suite` out of testcase id. When all tests have been migrated to the new ztest framework, maybe we can leverage the test summary (https://github.com/zephyrproject-rtos/zephyr/pull/48499) to collect the result.  That'll be simpler and more accurate.